File: backend/scripts/fill-ocr-gaps.py
#!/usr/bin/env python
"""
OCR-based gap filler for image-PDF MoEX question papers.

Targets:
  tcm1 102-2 tcm_basic_2 missing Q: 6,40,50,51,59,60,65,77,78,79,80
  nursing 108-1 psych_community entire paper (~80 Q)
"""
import os, sys, json, re, io, argparse, urllib3
os.environ['FLAGS_use_mkldnn']='0'
os.environ['FLAGS_enable_pir_in_executor']='0'
import requests
import fitz  # PyMuPDF
from pathlib import Path

urllib3.disable_warnings()
os.environ.setdefault('PYTHONIOENCODING','utf-8')

ROOT = Path(__file__).resolve().parent.parent
TMP  = ROOT / '_tmp'
PDFS = TMP / 'ocr_pdfs'
IMGS = TMP / 'ocr_imgs'
PDFS.mkdir(parents=True, exist_ok=True)
IMGS.mkdir(parents=True, exist_ok=True)

BASE = 'https://wwwq.moex.gov.tw/exam/wHandExamQandA_File.ashx'
UA = 'Mozilla/5.0'

def fetch_pdf(t, code, c, s, name):
    out = PDFS / f'{name}_{t}.pdf'
    if out.exists() and out.stat().st_size > 1000:
        return out
    url = f'{BASE}?t={t}&code={code}&c={c}&s={s}&q=1'
    r = requests.get(url, verify=False, timeout=60,
                     headers={'User-Agent': UA, 'Referer': 'https://wwwq.moex.gov.tw/'})
    if r.status_code != 200 or not r.content.startswith(b'%PDF'):
        print(f'  [fetch {t}] FAIL status={r.status_code} ct={r.headers.get("content-type")}')
        return None
    out.write_bytes(r.content)
    print(f'  [fetch {t}] saved {len(r.content)} bytes -> {out.name}')
    return out

def pdf_to_images(pdf_path, dpi=300):
    """Rasterize each page to PNG, return list of image paths."""
    doc = fitz.open(pdf_path)
    outs = []
    for i, page in enumerate(doc):
        pix = page.get_pixmap(dpi=dpi)
        out = IMGS / f'{pdf_path.stem}_p{i+1}.png'
        pix.save(str(out))
        outs.append(out)
    doc.close()
    return outs

_ocr = None
def ocr_engine():
    global _ocr
    if _ocr is None:
        from paddleocr import PaddleOCR
        # Try chinese_cht; fall back to ch
        # paddleocr 3.x API: use_textline_orientation; disable mkldnn to avoid IR bug
        kwargs = dict(use_textline_orientation=False, enable_mkldnn=False)
        try:
            _ocr = PaddleOCR(lang='chinese_cht', **kwargs)
        except Exception as e:
            print('  chinese_cht failed, falling back to ch:', e)
            _ocr = PaddleOCR(lang='ch', **kwargs)
    return _ocr

def ocr_image(img_path):
    """Return list of (text, bbox_center_y, bbox_min_x, bbox_max_x, bbox_min_y, bbox_max_y)."""
    eng = ocr_engine()
    # paddleocr 3.x API
    try:
        result = eng.predict(str(img_path))
        # result is list of dict per image
        items = []
        for page in result:
            texts = page.get('rec_texts', []) or []
            polys = page.get('rec_polys', []) or page.get('dt_polys', [])
            for t, poly in zip(texts, polys):
                xs = [p[0] for p in poly]; ys = [p[1] for p in poly]
                items.append({
                    'text': t, 'cx': sum(xs)/len(xs), 'cy': sum(ys)/len(ys),
                    'xmin': min(xs), 'xmax': max(xs), 'ymin': min(ys), 'ymax': max(ys),
                })
        return items
    except Exception as e:
        # fallback to 2.x API
        result = eng.ocr(str(img_path), cls=True)
        items = []
        if result and result[0]:
            for line in result[0]:
                poly, (text, _conf) = line
                xs=[p[0] for p in poly]; ys=[p[1] for p in poly]
                items.append({
                    'text': text, 'cx': sum(xs)/len(xs), 'cy': sum(ys)/len(ys),
                    'xmin':min(xs),'xmax':max(xs),'ymin':min(ys),'ymax':max(ys),
                })
        return items

def sort_reading_order(items, page_width):
    """Sort items left-column-first then right-column, top-to-bottom. MoEX papers are usually 2-column."""
    mid = page_width / 2
    left  = sorted([x for x in items if x['cx'] < mid], key=lambda x: x['cy'])
    right = sorted([x for x in items if x['cx'] >= mid], key=lambda x: x['cy'])
    return left + right

def cluster_lines(items, ytol=12):
    """Group OCR boxes into lines by y proximity within same column half."""
    # sort by y then x
    items = sorted(items, key=lambda x:(x['cy'], x['xmin']))
    lines = []
    cur = []
    cur_y = None
    for it in items:
        if cur_y is None or abs(it['cy']-cur_y) <= ytol:
            cur.append(it)
            cur_y = it['cy'] if cur_y is None else (cur_y+it['cy'])/2
        else:
            lines.append(cur); cur=[it]; cur_y=it['cy']
    if cur: lines.append(cur)
    # each line: sort by x, join text
    out = []
    for line in lines:
        line = sorted(line, key=lambda x:x['xmin'])
        out.append({
            'text': ''.join(w['text'] for w in line),
            'cy': sum(w['cy'] for w in line)/len(line),
            'xmin': min(w['xmin'] for w in line),
            'xmax': max(w['xmax'] for w in line),
        })
    return out

def parse_questions(all_lines_per_page, page_widths):
    """Input: list of lines per page. Each page processed in 2-column reading order.
    Returns list of {number, question, options:{A,B,C,D}}."""
    # Flatten into reading order: for each page, split into left/right columns, sort by y.
    ordered = []
    for plines, pw in zip(all_lines_per_page, page_widths):
        mid = pw/2
        left  = sorted([l for l in plines if (l['xmin']+l['xmax'])/2 < mid], key=lambda x:x['cy'])
        right = sorted([l for l in plines if (l['xmin']+l['xmax'])/2 >= mid], key=lambda x:x['cy'])
        ordered.extend(left); ordered.extend(right)

    # Concatenate text with newlines for regex parsing
    full = '\n'.join(l['text'] for l in ordered)
    # Quick normalise
    full = full.replace('（','(').replace('）',')')
    # Split by question starts: number followed by . or 、 at line start, 1-80
    # Pattern: (\d{1,2})\.  at start of a line
    return full, ordered

QNUM_RE = re.compile(r'^\s*(\d{1,2})[\.．、\s]\s*(.*)$')
OPT_RE  = re.compile(r'^\s*\(?([ABCD])\)?[\.．、\s]?\s*(.*)$')

def extract_qas(lines):
    """Extract question dicts from ordered lines."""
    qs = []
    cur = None
    mode = 'q'  # 'q' means collecting stem, 'opt' collecting options
    for l in lines:
        text = l['text'].strip()
        if not text: continue
        m = QNUM_RE.match(text)
        mo = OPT_RE.match(text)
        if m and (cur is None or len(cur.get('options',{}))>=1 or int(m.group(1))==cur['number']+1 or int(m.group(1))<=80):
            # Start new question if previous has options or numbers sequence fits
            if cur is not None:
                qs.append(cur)
            num = int(m.group(1))
            stem = m.group(2).strip()
            cur = {'number': num, 'question': stem, 'options': {}}
            mode = 'q'
            continue
        if mo and cur is not None:
            letter = mo.group(1)
            val = mo.group(2).strip()
            cur['options'][letter] = val
            cur['_last_opt'] = letter
            mode = 'opt'
            continue
        if cur is None: continue
        # continuation line
        if mode == 'q':
            cur['question'] += text
        elif mode == 'opt' and cur.get('_last_opt'):
            cur['options'][cur['_last_opt']] += text
    if cur is not None: qs.append(cur)
    # cleanup
    for q in qs: q.pop('_last_opt', None)
    return qs

def parse_answer_pdf(pdf_path):
    """Extract answer map {num: 'A'}.
    MoEX answer PDFs use column layout: all numbers then all letters in reading order,
    with '題號'/'答案' header markers. Parse by collecting contiguous number runs and
    letter runs, then zipping.
    """
    doc = fitz.open(pdf_path)
    text = ''
    for p in doc: text += p.get_text() + '\n'
    doc.close()
    ans = {}
    # Try inline patterns first
    for m in re.finditer(r'(?<!\d)(\d{1,3})\s*[\.、]\s*([ABCDE])(?![A-Z])', text):
        n=int(m.group(1)); l=m.group(2)
        if 1<=n<=200 and n not in ans: ans[n]=l
    if len(ans)>=5:
        return ans
    # Fallback: collect all letter-only lines after first '答案' marker in order → index 1..N
    lines = [ln.strip() for ln in text.splitlines() if ln.strip()]
    started = False
    letters = []
    for ln in lines:
        if ln == '答案' or '答案：' in ln:
            started = True
            continue
        if not started: continue
        if re.fullmatch(r'[ABCDE#]', ln):
            letters.append(ln)
        # Ignore 題號/答案 section headers between blocks
    if letters:
        for i, l in enumerate(letters, start=1):
            if l in 'ABCDE': ans[i] = l
        return ans
    # Column layout: iterate lines, find each "題號...答案" block followed by
    # a list of numbers then a list of letters.
    lines = [ln.strip() for ln in text.splitlines() if ln.strip()]
    i = 0
    while i < len(lines):
        if '題號' in lines[i]:
            # find '答案' marker
            j = i + 1
            nums = []
            while j < len(lines) and lines[j] != '答案' and not lines[j].startswith('答案'):
                if re.fullmatch(r'\d{1,3}', lines[j]):
                    nums.append(int(lines[j]))
                    j += 1
                else:
                    break
            # advance past 答案 marker
            while j < len(lines) and ('答案' in lines[j]):
                j += 1
            letters = []
            while j < len(lines) and len(letters) < len(nums):
                if re.fullmatch(r'[ABCDE#]', lines[j]):
                    letters.append(lines[j])
                    j += 1
                else:
                    break
            for n, l in zip(nums, letters):
                if l in 'ABCDE' and 1 <= n <= 200 and n not in ans:
                    ans[n] = l
            i = j
        else:
            i += 1
    return ans

def load_bank(fp):
    return json.loads(Path(fp).read_text(encoding='utf-8'))

def save_bank(fp, data):
    Path(fp).write_text(json.dumps(data, ensure_ascii=False, indent=2), encoding='utf-8')

def merge_into_bank(bank_path, new_entries, exam_code):
    data = load_bank(bank_path)
    qs = data['questions']
    existing = {(q.get('exam_code'), q.get('subject_tag'), q.get('number')) for q in qs}
    added = 0
    for e in new_entries:
        key = (e['exam_code'], e['subject_tag'], e['number'])
        if key in existing:
            continue
        qs.append(e); added += 1
    data['total'] = len(qs)
    save_bank(bank_path, data)
    return added

def build_entry(roc_year, session, exam_code, subject, subject_tag, subject_name, number, question, options, answer):
    id_ = f"{exam_code}_{session_code(session)}_{number}"
    return {
        'id': id_,
        'roc_year': roc_year,
        'session': session,
        'exam_code': exam_code,
        'subject': subject,
        'subject_tag': subject_tag,
        'subject_name': subject_name,
        'stage_id': 0,
        'number': number,
        'question': question,
        'options': options,
        'answer': answer,
        'explanation': '',
    }

def session_code(session):
    return {'第一次':'0101','第二次':'0202'}.get(session,'0000')

def run_target(name, code, c, s, bank_path, roc_year, session, subject, subject_tag, subject_name, wanted_numbers=None, dry=False):
    print(f'\n=== {name} ===')
    qpdf = fetch_pdf('Q', code, c, s, name)
    if qpdf is None: return 0
    apdf = fetch_pdf('S', code, c, s, name) or fetch_pdf('A', code, c, s, name) or fetch_pdf('M', code, c, s, name)

    answers = parse_answer_pdf(apdf) if apdf else {}
    print(f'  answer PDF parsed: {len(answers)} answers')

    # rasterize
    imgs = pdf_to_images(qpdf, dpi=300)
    print(f'  pages: {len(imgs)}')

    # OCR
    all_lines = []
    page_widths = []
    for i, img in enumerate(imgs):
        items = ocr_image(img)
        # page width
        doc = fitz.open(qpdf); page = doc[i]
        pw = page.rect.width * 300/72  # convert to image px
        doc.close()
        lines = cluster_lines(items)
        all_lines.append(lines); page_widths.append(pw)
        print(f'    page {i+1}: {len(items)} boxes, {len(lines)} lines')

    # Build ordered lines across pages with column sorting
    ordered = []
    for plines, pw in zip(all_lines, page_widths):
        mid = pw/2
        left  = sorted([l for l in plines if (l['xmin']+l['xmax'])/2 < mid], key=lambda x:x['cy'])
        right = sorted([l for l in plines if (l['xmin']+l['xmax'])/2 >= mid], key=lambda x:x['cy'])
        ordered.extend(left); ordered.extend(right)

    # Debug: dump first 40 ordered lines
    print('  === first 40 ordered lines ===')
    for l in ordered[:40]:
        print(f'    [{l["cy"]:.0f}/{l["xmin"]:.0f}] {l["text"][:90]}')
    qs = extract_qas(ordered)
    print(f'  parsed {len(qs)} questions')

    # Filter to wanted set if provided
    if wanted_numbers is not None:
        qs = [q for q in qs if q['number'] in wanted_numbers]
        print(f'  after filter: {len(qs)}')

    # Build entries
    entries = []
    skipped = 0
    for q in qs:
        num = q['number']
        opts = q['options']
        if not all(k in opts for k in 'ABCD'):
            print(f'    skip Q{num}: missing options {sorted(opts.keys())}')
            skipped += 1
            continue
        ans = answers.get(num)
        if not ans:
            print(f'    skip Q{num}: no answer')
            skipped += 1
            continue
        entries.append(build_entry(
            roc_year, session, code, subject, subject_tag, subject_name,
            num, q['question'].strip(), {k:opts[k].strip() for k in 'ABCD'}, ans))

    # Spot-check first 3
    for e in entries[:3]:
        print(f'  Q{e["number"]}: {e["question"][:60]}')
        for k in 'ABCD': print(f'     ({k}) {e["options"][k][:50]}')
        print(f'     ANS: {e["answer"]}')

    if dry:
        print(f'  DRY-RUN: would add {len(entries)} entries ({skipped} skipped)')
        return 0
    added = merge_into_bank(bank_path, entries, code)
    print(f'  merged into {bank_path}: +{added} ({skipped} skipped)')
    return added

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument('--target', choices=['tcm1','nursing','all'], default='all')
    ap.add_argument('--dry', action='store_true')
    args = ap.parse_args()

    total = 0
    if args.target in ('tcm1','all'):
        total += run_target(
            name='tcm1_102-2_basic2',
            code='102110', c='103', s='0202',
            bank_path=ROOT/'questions-tcm1.json',
            roc_year='102', session='第二次',
            subject='中醫基礎醫學(二)', subject_tag='tcm_basic_2',
            subject_name='中醫基礎醫學(二)',
            wanted_numbers={6,40,50,51,59,60,65,77,78,79,80},
            dry=args.dry,
        )
    if args.target in ('nursing','all'):
        total += run_target(
            name='nursing_108-1_psych',
            code='108020', c='106', s='0505',
            bank_path=ROOT/'questions-nursing.json',
            roc_year='108', session='第一次',
            subject='精神科與社區衛生護理學', subject_tag='psych_community',
            subject_name='精神科與社區衛生護理學',
            wanted_numbers=None,
            dry=args.dry,
        )
    print(f'\nDONE. total added: {total}')

if __name__ == '__main__':
    main()

File: backend/scripts/fill-ocr-gaps.js
#!/usr/bin/env node
/**
 * OCR gap-fill driver.
 * Reads per-page OCR JSON from _tmp/ocr_json/<prefix>_p{N}.json,
 * parses questions, matches answers from the answer PDF (via pdf-parse),
 * and merges into the target questions bank.
 *
 * Usage:
 *   node scripts/fill-ocr-gaps.js --target tcm1 [--dry]
 *   node scripts/fill-ocr-gaps.js --target nursing [--dry]
 */
const fs = require('fs');
const path = require('path');
const pdfParse = require('pdf-parse');

const ROOT = path.resolve(__dirname, '..');
const TMP = path.join(ROOT, '_tmp');
const OCR_JSON = path.join(TMP, 'ocr_json');
const PDFS = path.join(TMP, 'ocr_pdfs');

const TARGETS = {
  tcm1: {
    prefix: 'tcm1_p',
    pages: 10,
    qPdf: path.join(PDFS, 'tcm1_102-2_basic2_Q.pdf'),
    sPdf: path.join(PDFS, 'tcm1_102-2_basic2_S.pdf'),
    bank: path.join(ROOT, 'questions-tcm1.json'),
    wanted: new Set([6, 40, 50, 51, 59, 60, 65, 77, 78, 79, 80]),
    meta: {
      roc_year: '102', session: '第二次', exam_code: '102110',
      subject: '中醫基礎醫學(二)', subject_tag: 'tcm_basic_2',
      subject_name: '中醫基礎醫學(二)',
    },
  },
  nursing: {
    prefix: 'nursing_p',
    pages: 8,
    qPdf: path.join(PDFS, 'nursing_108-1_psych_Q.pdf'),
    sPdf: path.join(PDFS, 'nursing_108-1_psych_S.pdf'),
    bank: path.join(ROOT, 'questions-nursing.json'),
    wanted: null, // all
    meta: {
      roc_year: '108', session: '第一次', exam_code: '108020',
      subject: '精神科與社區衛生護理學', subject_tag: 'psych_community',
      subject_name: '精神科與社區衛生護理學',
    },
  },
};

// ----- OCR JSON -> lines (cluster boxes into lines, per page, column-aware) -----
function loadPage(p) {
  return JSON.parse(fs.readFileSync(p, 'utf-8'));
}
function clusterLines(items, ytol = 18) {
  // sort by cy, then xmin
  const arr = [...items].sort((a, b) => a.cy - b.cy || a.xmin - b.xmin);
  const lines = [];
  let cur = [], curY = null;
  for (const it of arr) {
    if (curY == null || Math.abs(it.cy - curY) <= ytol) {
      cur.push(it);
      curY = curY == null ? it.cy : (curY + it.cy) / 2;
    } else {
      lines.push(cur);
      cur = [it]; curY = it.cy;
    }
  }
  if (cur.length) lines.push(cur);
  return lines.map(line => {
    line.sort((a, b) => a.xmin - b.xmin);
    return {
      text: line.map(w => w.text).join(''),
      parts: line,
      cy: line.reduce((s, w) => s + w.cy, 0) / line.length,
      xmin: Math.min(...line.map(w => w.xmin)),
      xmax: Math.max(...line.map(w => w.xmax)),
    };
  });
}

// Split a "line" into per-column segments based on x ranges.
// MoEX papers are single-column for stems, but options usually sit on one row (A|B|C|D) in 4 columns.
// Better approach: don't cluster across full page width. Cluster using per-box y but split boxes by column based on option anchors.
// Simpler: detect option rows by presence of (A)/(B)/(C)/(D) markers in the same clustered line.

function splitOptionRow(line) {
  // Returns { A, B, C, D } if this line has multiple option markers; else null.
  const opts = {};
  const parts = line.parts;
  // Find option anchors in parts
  const anchors = [];
  parts.forEach((p, i) => {
    const m = p.text.match(/^\(?([ABCD])\)?[.\s、．]?(.*)$/);
    if (m) anchors.push({ idx: i, letter: m[1], rest: m[2] || '' });
  });
  // A valid option row has >=2 distinct letters
  const letters = [...new Set(anchors.map(a => a.letter))];
  if (letters.length < 2) return null;
  // Walk anchors: between each anchor and next, concat parts' text
  for (let k = 0; k < anchors.length; k++) {
    const a = anchors[k];
    const end = k + 1 < anchors.length ? anchors[k + 1].idx : parts.length;
    let txt = a.rest;
    for (let j = a.idx + 1; j < end; j++) txt += parts[j].text;
    opts[a.letter] = (opts[a.letter] || '') + txt;
  }
  return opts;
}

// Detect question-number anchor: text starts with plain "N" digit alone, x at far left (xmin small).
function isQnumAnchor(line) {
  if (!line.parts.length) return null;
  const first = line.parts[0];
  // accept if first box is pure digits 1-2 chars and at small xmin (< 200 px for 300dpi)
  if (!/^\d{1,2}$/.test(first.text.trim())) return null;
  if (first.xmin > 260) return null;
  const n = parseInt(first.text.trim(), 10);
  if (n < 1 || n > 100) return null;
  return n;
}

// Extract questions from page lines.
function parsePages(target) {
  const allLines = [];
  for (let i = 1; i <= target.pages; i++) {
    const fp = path.join(OCR_JSON, `${target.prefix}${i}.json`);
    if (!fs.existsSync(fp)) { console.error('MISSING', fp); continue; }
    const items = loadPage(fp);
    // drop tiny noise boxes? no — keep
    const lines = clusterLines(items);
    // filter out header (cy < 1000 on page 1 only, contains page number etc)
    for (const l of lines) allLines.push({ page: i, ...l });
  }
  return allLines;
}

function extractQuestions(allLines) {
  const qs = new Map(); // number -> { question, options }
  let cur = null;
  let mode = 'pre';
  for (const l of allLines) {
    const text = l.text.trim();
    if (!text) continue;
    // try question anchor
    const qn = isQnumAnchor(l);
    // detect option row
    const opts = splitOptionRow(l);
    if (qn != null && !opts) {
      // close previous
      if (cur && !qs.has(cur.number)) qs.set(cur.number, cur);
      // stem is rest of line after first box
      const stem = l.parts.slice(1).map(p => p.text).join('').trim();
      cur = { number: qn, question: stem, options: {} };
      mode = 'q';
      continue;
    }
    if (opts) {
      if (cur) {
        Object.assign(cur.options, opts);
        mode = 'opt';
        cur._lastOpt = 'D' in opts ? 'D' : Object.keys(opts).sort().pop();
      }
      continue;
    }
    // continuation
    if (!cur) continue;
    if (mode === 'q') cur.question += text;
    else if (mode === 'opt' && cur._lastOpt) cur.options[cur._lastOpt] += text;
  }
  if (cur && !qs.has(cur.number)) qs.set(cur.number, cur);
  return qs;
}

// ----- Answer PDF parsing -----
async function parseAnswerPdf(pdfPath) {
  const buf = fs.readFileSync(pdfPath);
  const { text } = await pdfParse(buf);
  const ans = {};
  // Strategy 1: inline "1.A" or "1、A"
  for (const m of text.matchAll(/(?<!\d)(\d{1,3})\s*[.、．]\s*([A-E#])(?![A-Z])/g)) {
    const n = parseInt(m[1], 10);
    if (n >= 1 && n <= 200 && !ans[n]) ans[n] = m[2];
  }
  if (Object.keys(ans).length >= 20) return ans;
  // Strategy 2: "答案" followed by sequence of letters (half or full width)
  const fullWidthMap = { 'Ａ': 'A', 'Ｂ': 'B', 'Ｃ': 'C', 'Ｄ': 'D', 'Ｅ': 'E', '＃': '#' };
  // Collect all "答案" blocks
  const blocks = text.split(/答\s*案/);
  let out = [];
  for (let i = 1; i < blocks.length; i++) {
    const b = blocks[i];
    // take chars until next non-letter cluster ends
    const cleaned = b.split('').map(c => fullWidthMap[c] || c).join('');
    const letters = cleaned.match(/[ABCDE#]/g) || [];
    if (letters.length >= 10) out = out.concat(letters);
  }
  if (out.length) {
    out.forEach((l, i) => { if (!ans[i + 1] && 'ABCDE'.includes(l)) ans[i + 1] = l; });
  }
  return ans;
}

// ----- bank merge -----
function sessionCode(sess) {
  return { '第一次': '0101', '第二次': '0202' }[sess] || '0000';
}
function buildEntry(meta, num, question, options, answer) {
  return {
    id: `${meta.exam_code}_${sessionCode(meta.session)}_${num}`,
    roc_year: meta.roc_year,
    session: meta.session,
    exam_code: meta.exam_code,
    subject: meta.subject,
    subject_tag: meta.subject_tag,
    subject_name: meta.subject_name,
    stage_id: 0,
    number: num,
    question,
    options,
    answer,
    explanation: '',
  };
}

function merge(bankPath, entries) {
  const data = JSON.parse(fs.readFileSync(bankPath, 'utf-8'));
  const existing = new Set(data.questions.map(q => `${q.exam_code}|${q.subject_tag}|${q.number}`));
  let added = 0;
  for (const e of entries) {
    const k = `${e.exam_code}|${e.subject_tag}|${e.number}`;
    if (existing.has(k)) continue;
    data.questions.push(e); added++;
  }
  data.total = data.questions.length;
  fs.writeFileSync(bankPath, JSON.stringify(data, null, 2), 'utf-8');
  return added;
}

async function runTarget(name, target, dry) {
  console.log(`\n=== ${name} ===`);
  const lines = parsePages(target);
  console.log(`  lines: ${lines.length}`);
  const qsMap = extractQuestions(lines);
  console.log(`  parsed questions: ${qsMap.size} (numbers: ${[...qsMap.keys()].sort((a, b) => a - b).join(',')})`);

  // answers
  let answers = {};
  if (fs.existsSync(target.sPdf)) {
    try {
      answers = await parseAnswerPdf(target.sPdf);
      console.log(`  answers parsed: ${Object.keys(answers).length}`);
    } catch (e) { console.error('  answer parse error:', e.message); }
  } else {
    console.warn('  NO answer PDF at', target.sPdf);
  }

  // filter
  const entries = [];
  const skipped = [];
  for (const [num, q] of qsMap) {
    if (target.wanted && !target.wanted.has(num)) continue;
    const opts = q.options;
    if (!['A', 'B', 'C', 'D'].every(k => opts[k] != null && opts[k].length > 0)) {
      skipped.push({ num, reason: 'missing_opts', have: Object.keys(opts) });
      continue;
    }
    const ans = answers[num];
    if (!ans) { skipped.push({ num, reason: 'no_answer' }); continue; }
    if (!q.question.trim()) { skipped.push({ num, reason: 'no_stem' }); continue; }
    entries.push(buildEntry(target.meta, num,
      q.question.trim(),
      { A: opts.A.trim(), B: opts.B.trim(), C: opts.C.trim(), D: opts.D.trim() },
      ans));
  }

  console.log(`  entries built: ${entries.length}`);
  if (skipped.length) console.log(`  skipped:`, skipped.slice(0, 20));
  // spot-check
  for (const e of entries.slice(0, 3)) {
    console.log(`  --- Q${e.number} ANS=${e.answer} ---`);
    console.log(`    stem: ${e.question.slice(0, 100)}`);
    for (const k of 'ABCD') console.log(`    (${k}) ${e.options[k].slice(0, 80)}`);
  }

  if (dry) {
    console.log(`  DRY: would add ${entries.length}`);
    return 0;
  }
  const added = merge(target.bank, entries);
  console.log(`  merged into ${path.basename(target.bank)}: +${added}`);
  return added;
}

(async () => {
  const args = process.argv.slice(2);
  const dry = args.includes('--dry');
  const tIdx = args.indexOf('--target');
  const which = tIdx >= 0 ? args[tIdx + 1] : 'all';
  let total = 0;
  if (which === 'all' || which === 'tcm1') total += await runTarget('tcm1_102-2_basic2', TARGETS.tcm1, dry);
  if (which === 'all' || which === 'nursing') total += await runTarget('nursing_108-1_psych', TARGETS.nursing, dry);
  console.log(`\nDONE. total added: ${total}`);
})();
